File: trust-issues-dash/src/app/lib/prediction_service.ts
import { doc, setDoc, serverTimestamp, updateDoc } from "firebase/firestore";
import { db } from "./firebase"; // Ensure this points to your firebase config

// ==========================================
// 1. CONFIGURATION & INTERFACES
// ==========================================

// Add these to your .env.local file:
// NEXT_PUBLIC_RAILWAY_API_URL=https://your-railway-app.up.railway.app
// NEXT_PUBLIC_RAILWAY_API_KEY=your_secret_key
const API_URL = process.env.NEXT_PUBLIC_RAILWAY_API_URL?.replace(/\/$/, '');
const API_KEY = process.env.NEXT_PUBLIC_RAILWAY_API_KEY;

export interface PredictionResponse {
    prediction: string; // e.g., "SPAM" or "HAM"
    [key: string]: any;
}

export interface ExplanationResponse {
    why_spam: string;
    spam_type: string;
    detection_tips: string;
    error?: string | null;
}

// Helper to generate readable timestamp IDs (e.g., "2026-04-21T18_14_11Z")
const generateTimestampId = () => new Date().toISOString().replace(/[:.]/g, '_');

// ==========================================
// 2. CORE FETCH WRAPPER
// ==========================================

async function railwayFetch<T>(endpoint: string, body: any): Promise<T> {
    // Grab them fresh right here
    const apiUrl = process.env.NEXT_PUBLIC_RAILWAY_API_URL?.replace(/\/$/, '');
    const apiKey = process.env.NEXT_PUBLIC_RAILWAY_API_KEY;

    if (!apiUrl || !apiKey) {
        console.error("DEBUG ENV:", { apiUrl, apiKey }); // This will print to your browser console
        throw new Error("Missing Railway API environment variables.");
    }

    const response = await fetch(`${apiUrl}${endpoint}`, {
        method: "POST",
        headers: {
            "Content-Type": "application/json",
            "Trust-issue-API-Key": apiKey,
        },
        body: JSON.stringify(body),
    });

    if (!response.ok) {
        throw new Error(`Railway API Error: ${response.status} ${response.statusText}`);
    }

    return response.json();
}

// ==========================================
// 3. SMS & EMAIL PREDICTION SERVICES
// ==========================================

export const checkSms = async (uid: string, sender: string, text: string) => {
    // 1. Ask the AI Backend
    const result = await railwayFetch<PredictionResponse>("/api/v1/predict/sms", { text });
    console.log("SMS API Response:", result);

    // Extract the nested prediction ("SPAM" or "HAM")
    const finalPrediction = result?.data?.prediction || "UNKNOWN";

    // 2. Format the document ID
    const docId = generateTimestampId();
    const docRef = doc(db, "users", uid, "sms_checks", docId);

    // 3. Save to Firebase exactly as defined in the schema
    await setDoc(docRef, {
        message: text,
        prediction: finalPrediction,
        pushed_to_community: false,
        sender: sender,
        timestamp: serverTimestamp(),
    });

    return { docId, prediction: finalPrediction };
};

export const checkEmail = async (uid: string, sender: string, subject: string, body: string) => {
    // 1. Ask the AI Backend
    const result = await railwayFetch<PredictionResponse>("/api/v1/predict/email", { subject, body });
    console.log("Email API Response:", result);

    // Extract the nested prediction ("SPAM" or "HAM")
    const finalPrediction = result?.data?.prediction || "UNKNOWN";

    // 2. Format the document ID
    const docId = generateTimestampId();
    const docRef = doc(db, "users", uid, "email_checks", docId);

    // 3. Save to Firebase
    await setDoc(docRef, {
        message: body,
        subject: subject,
        prediction: finalPrediction,
        pushed_to_community: false,
        sender: sender,
        timestamp: serverTimestamp(),
    });

    return { docId, prediction: finalPrediction };
};

// ==========================================
// 4. AI EXPLANATION & SYNC SERVICE
// ==========================================

export const explainMessage = async (
    uid: string,
    source: "email" | "sms",
    originalDocId: string, // Needed so we can update the original doc!
    subject: string | null,
    body: string,
    ml_model_output: string
) => {
    // 1. Ask the AI Backend for the breakdown
    const explanation = await railwayFetch<ExplanationResponse>("/api/v1/explain", {
        source,
        subject,
        body,
        ml_model_output,
    });

    // 2. Format a clean AI explanation string based on the response
    const combinedExplanation = `${explanation.why_spam} ${explanation.detection_tips}`;

    // 3. Save the specific AI Ask into the `ai_asks` subcollection
    const aiDocId = generateTimestampId();
    const aiDocRef = doc(db, "users", uid, "ai_asks", aiDocId);

    await setDoc(aiDocRef, {
        ai_explanation: combinedExplanation,
        original_text: body,
        spam_type: explanation.spam_type,
        timestamp: serverTimestamp(),
    });

    // 4. Update the original SMS or Email document with the new findings
    // (The prompt mentioned updating the classification in the original doc)
    const collectionName = source === "email" ? "email_checks" : "sms_checks";
    const originalDocRef = doc(db, "users", uid, collectionName, originalDocId);

    await updateDoc(originalDocRef, {
        ai_explanation_ref: aiDocId, // Handly link to the explanation
        detailed_spam_type: explanation.spam_type, // Inject the exact type
    });

    return explanation;
};

File: lib_from_moob_back_up/services/prediction_service.dart
import 'dart:convert';
import 'dart:developer' as developer;
import 'package:flutter_dotenv/flutter_dotenv.dart';
import 'package:http/http.dart' as http;
import 'package:cloud_firestore/cloud_firestore.dart';

class PredictionService {
  final FirebaseFirestore _firestore = FirebaseFirestore.instance;

  // ==========================================
  // 1. CONFIGURATION
  // ==========================================
  
  // TODO: Replace with your actual Railway URL and Key 
  // (Alternatively, use the flutter_dotenv package to hide these)
  static final String _apiUrl = dotenv.env['RAILWAY_API_URL'] ?? 'MISSING_URL';
  static final String _apiKey = dotenv.env['RAILWAY_API_KEY'] ?? 'MISSING_KEY';

  // Helper to generate readable timestamp IDs (e.g., "2026-04-21T18_14_11Z")
  String _generateTimestampId() {
    return DateTime.now().toUtc().toIso8601String().replaceAll(RegExp(r'[:.]'), '_');
  }

  // ==========================================
  // 2. CORE FETCH WRAPPER
  // ==========================================
  
  Future<Map<String, dynamic>> _railwayFetch(String endpoint, Map<String, dynamic> body) async {
    try {
      final uri = Uri.parse('$_apiUrl$endpoint');
      
      final response = await http.post(
        uri,
        headers: {
          'Content-Type': 'application/json',
          'Trust-issue-API-Key': _apiKey,
        },
        body: jsonEncode(body),
      );

      if (response.statusCode >= 200 && response.statusCode < 300) {
        return jsonDecode(response.body);
      } else {
        throw Exception('Railway API Error: ${response.statusCode} ${response.reasonPhrase}');
      }
    } catch (e, stackTrace) {
      developer.log('API Fetch Error -> $endpoint', error: e, stackTrace: stackTrace, name: 'PredictionService');
      rethrow;
    }
  }

  // ==========================================
  // 3. SMS & EMAIL PREDICTION SERVICES
  // ==========================================

  Future<Map<String, dynamic>> checkSms({
    required String uid,
    required String sender,
    required String text,
  }) async {
    developer.log('Analyzing SMS from $sender...', name: 'PredictionService');
    
    // 1. Ask the AI Backend
    final result = await _railwayFetch('/api/v1/predict/sms', {'text': text});
    
    // Extract the nested prediction ("SPAM" or "HAM")
    final String finalPrediction = result['data']?['prediction'] ?? 'UNKNOWN';

    // 2. Format the document ID
    final docId = _generateTimestampId();
    final docRef = _firestore.collection('users').doc(uid).collection('sms_checks').doc(docId);

    // 3. Save to Firebase exactly as defined in the web schema
    await docRef.set({
      'message': text,
      'prediction': finalPrediction,
      'pushed_to_community': false,
      'sender': sender,
      'timestamp': FieldValue.serverTimestamp(),
    });

    developer.log('SMS Logged: $finalPrediction', name: 'PredictionService');
    return {'docId': docId, 'prediction': finalPrediction};
  }

  Future<Map<String, dynamic>> checkEmail({
    required String uid,
    required String sender,
    required String subject,
    required String body,
  }) async {
    developer.log('Analyzing Email from $sender...', name: 'PredictionService');

    // 1. Ask the AI Backend
    final result = await _railwayFetch('/api/v1/predict/email', {
      'subject': subject,
      'body': body,
    });
    
    // Extract the nested prediction
    final String finalPrediction = result['data']?['prediction'] ?? 'UNKNOWN';

    // 2. Format the document ID
    final docId = _generateTimestampId();
    final docRef = _firestore.collection('users').doc(uid).collection('email_checks').doc(docId);

    // 3. Save to Firebase
    await docRef.set({
      'message': body,
      'subject': subject,
      'prediction': finalPrediction,
      'pushed_to_community': false,
      'sender': sender,
      'timestamp': FieldValue.serverTimestamp(),
    });

    developer.log('Email Logged: $finalPrediction', name: 'PredictionService');
    return {'docId': docId, 'prediction': finalPrediction};
  }

  // ==========================================
  // 4. AI EXPLANATION & SYNC SERVICE
  // ==========================================

  Future<Map<String, dynamic>> explainMessage({
    required String uid,
    required String source, // "email" or "sms"
    required String originalDocId,
    String? subject,
    required String body,
    required String mlModelOutput,
  }) async {
    developer.log('Fetching AI Explanation for $source...', name: 'PredictionService');

    // 1. Ask the AI Backend for the breakdown
    final explanation = await _railwayFetch('/api/v1/explain', {
      'source': source,
      'subject': subject, // Can be null for SMS
      'body': body,
      'ml_model_output': mlModelOutput,
    });

    // 2. Format a clean AI explanation string
    final String whySpam = explanation['why_spam'] ?? '';
    final String tips = explanation['detection_tips'] ?? '';
    final String combinedExplanation = '$whySpam $tips'.trim();
    final String spamType = explanation['spam_type'] ?? 'Unknown Type';

    // 3. Save the specific AI Ask into the `ai_asks` subcollection
    final aiDocId = _generateTimestampId();
    final aiDocRef = _firestore.collection('users').doc(uid).collection('ai_asks').doc(aiDocId);

    await aiDocRef.set({
      'ai_explanation': combinedExplanation,
      'original_text': body,
      'spam_type': spamType,
      'timestamp': FieldValue.serverTimestamp(),
    });

    // 4. Update the original SMS or Email document with the new findings
    final collectionName = source == 'email' ? 'email_checks' : 'sms_checks';
    final originalDocRef = _firestore.collection('users').doc(uid).collection(collectionName).doc(originalDocId);

    await originalDocRef.update({
      'ai_explanation_ref': aiDocId,
      'detailed_spam_type': spamType,
    });

    developer.log('Explanation Synced Successfully.', name: 'PredictionService');
    return explanation;
  }
}